### PR TITLE
e_devcrypto:  e_devcrypto: Alleviate problems with forks

### PR DESCRIPTION
**Pull request by cotequeiroz**
This PR is an attempt to alleviate problems described in #8430.  It's far from fixing it, so I did not write that in the commit message.  The C program I wrote there will still fail with this applied, but I plan to close the issue if this is merged nonetheless, as that case is fictional, openssh is real.  One can always reopen or open another issue if something else is found to be broken.

I'm covering the most common cases I observed here:
 - Errors in ECB ciphers used in `drbg_ctr_init()` across forks.  I'm just ignoring an error closing a previous open session in `cipher_init`, so init does not fail because of it.  With ECB there is no partial state data to be copied, so this fixes it.
 - The HMAC digests done across forks that completely break openssh
The previous `cryptodev` engine worked by storing digest input data, and then running it though /dev/crypto only when `DigestFinal` was called.  This worked with forks, as all that needed to be duplicated was the temporary storage.

Then with the new engine, we feed data to cryptodev at every `DigestUpdate` call.  What I'm doing here is to keep a small data cache that is only going to be fed to cryptodev when its size is > 8k, and also deferring cryptodev initialization until we are actually sending data to it.  Since the common case observed is the HMAC inner key, by the time the process forks, the cryptodev session should not be open, and the only thing we're copying is the input cache data.  In order to avoid duplicating a large chunk of memory, if the new data arriving is larger than 256K, then the current cache data is sent to cryptodev, and then another call is made to send the new data.  Both of these parameters can be changed in `CFLAGS`.  

I've also added an `#ifndef` check to `DEVCRYPTO_DEFAULT_USE_SOFTDRIVERS`, so it could be changed in `CFLAGS` as well--I use that to allow testing in a x86_64 without hw-crypto, rather than using a slower openwrt router every time.

I have tested this with `make test` on x86_64, and with openssh-8.0 in a Linksys WRT3200ACM running openwrt master, confirming that it does not run with the current engine, it exits with
```
# /usr/sbin/sshd -ddd
debug2: load_server_config: filename /etc/ssh/sshd_config
debug2: load_server_config: done config len = 272
debug2: parse_server_config: config /etc/ssh/sshd_config len 272
debug3: /etc/ssh/sshd_config:13 setting Port 2200
debug3: /etc/ssh/sshd_config:19 setting HostKey /etc/ssh/ssh_host_ecdsa_key
debug3: /etc/ssh/sshd_config:20 setting HostKey /etc/ssh/ssh_host_ed25519_key
debug3: /etc/ssh/sshd_config:41 setting AuthorizedKeysFile .ssh/authorized_keys
debug3: /etc/ssh/sshd_config:109 setting Subsystem sftp /usr/lib/sftp-server
debug1: sshd version OpenSSH_8.0, OpenSSL 1.1.1c  28 May 2019
sshkey_fingerprint failed
```
After this change, sshd works as expected.  BTW, openwrt has had 1dea0bf applied for a while now, so it has been tested to some extent.  Digests problem was dealt by disabling them by default.  People will try to turn them on and get into trouble eventually.

Ping @levitte

### CODE REVIEW

File: engines/e_devcrypto.c [hunk 1]
```diff
@@ -41,7 +41,9 @@ static int cfd = -1;
 #define DEVCRYPTO_USE_SOFTWARE        1 /* allow software drivers */
 #define DEVCRYPTO_REJECT_SOFTWARE     2 /* only disallow confirmed software drivers */
 
+#ifndef DEVCRYPTO_DEFAULT_USE_SOFTDRIVERS
 #define DEVCRYPTO_DEFAULT_USE_SOFTDRIVERS DEVCRYPTO_REJECT_SOFTWARE
```
Comment (levitte) [inline engines/e_devcrypto.c]: Coding style: Because of the guard, this needs to be indented by inserting a space between `#` and `define`

File: engines/e_devcrypto.c [hunk 2]
```diff
@@ -606,10 +607,23 @@ static void dump_cipher_info(void)
...
 #if defined(CIOCCPHASH) && defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL)
 #define IMPLEMENT_DIGEST
 
+#ifndef DIGEST_CACHE_SIZE
+#define DIGEST_CACHE_SIZE 8192
+#endif
+
+#ifndef DIGEST_CACHE_MAXSIZE
+#define DIGEST_CACHE_MAXSIZE 262144
+#endif
```
Comment (levitte) [inline engines/e_devcrypto.c]: Same kind of indents in this hunk.

File: engines/e_devcrypto.c [hunk 3]
```diff
@@ -772,29 +817,44 @@ static int digest_copy(EVP_MD_CTX *to, const EVP_MD_CTX *from)
...
-    if (!digest_init(to)) {
+    if (digest_from->inp_count > 0) {
+        digest_to->inp_data = OPENSSL_malloc(digest_from->inp_count);
+        if (digest_to->inp_data == NULL) {
+            fprintf(stderr, "devcrypto: malloc failed\n");
+            goto fail;
+        }
+        memcpy(digest_to->inp_data, digest_from->inp_data, digest_from->inp_count);
+    }
+    if (digest_from->init_called != 1)
```
Comment (dlegaultbbry) [inline engines/e_devcrypto.c]: This should be moved with digest_from == NULL above

File: engines/e_devcrypto.c [hunk 4]
```diff
@@ -698,70 +716,97 @@ static int digest_init(EVP_MD_CTX *ctx)
...
     if (digest_ctx == NULL)
         return 0;
 
-    if (EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT)) {
-        if (digest_op(digest_ctx, data, count, digest_ctx->digest_res, 0) >= 0)
-            return 1;
-    } else if (digest_op(digest_ctx, data, count, NULL, COP_FLAG_UPDATE) >= 0) {
-        return 1;
+    if (EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT))
+        return (digest_op(digest_ctx, data, count, digest_ctx->digest_res, 0) >= 0);
```
Comment (dlegaultbbry) [inline engines/e_devcrypto.c]: Shouldn't this use UPDATE | FINAL as flags?

**Comment by cotequeiroz** [comment]
> I'm a bit worried by those use cases... Why is an HMAC operation carried on across a fork?

I'll have to dig into openssh sources.  I imagine it forks every time it gets a new connection, but it would not be a place I would expect an open digest context, and turning debug on disables that fork anyway.  It also does privilege separation, where it may be doing it.  While running gdb, I remember seeing 4 inferiors.

The code is hard for me to read, as I remember it.  Looking from the openssl side was much easier for me.  It was a while ago, but I do remember it was awful to figure out what the problem was in the first place.

It goes through key exchange and cipher agreement, then here's the log where it fails:
```
debug3: send packet: type 31 [preauth]
debug3: send packet: type 21 [preauth]
debug2: set_newkeys: mode 1 [preauth]
ssh_dispatch_run_fatal: Connection from 192.168.144.20 port 50467: invalid argument [preauth]
debug1: do_cleanup [preauth]
debug1: monitor_read_log: child log fd closed
debug3: mm_request_receive entering
debug1: do_cleanup
debug1: Killing privsep child 6218
```
The invalid argument error comes from cryptodev, due to the session being closed by another process sharing the same session.  Since they are asynchronous, the error message would vary--another difficulty in finding out the problem.  Another possibility was:
```
 auth.info sshd[6380]: Corrupted MAC on input. [preauth]
auth.info sshd[6380]: ssh_dispatch_run_fatal: Connection from 192.168.144.20 port 57805: message authentication code incorrect [preauth]
```
where data would be fed to the same session by different processes, so the final digest will not be right.

File: engines/e_devcrypto.c (hunk 4 above)
Comment (cotequeiroz) [reply engines/e_devcrypto.c]: The original cryptodev implementation did not have update/final support.  It was added later, so the operation was initially done without the flags.  Documentation is sparse, so there's no guarantee, but looking at `main.c` in linux-cryptodev, it looks like it works with both flags as well.  I'm not sure it needs a reset, if so, a `COP_FLAG_RESET` would be needed.  When running without flags, it performs a reset as the first step, then it does update, and calls final.  I'd go with history and leave the flags zeroed.  Notice that this call was originally written (by me) without them, as was the case in the old cryptodev engine.

File: engines/e_devcrypto.c (hunk 3 above)
Comment (cotequeiroz) [reply engines/e_devcrypto.c]: If we move it up, it will return before copying the cached data.  `init_called` is referring to the cryptodev session having been initialized, and we're deferring this to the first update call--to accommodate the HMAC fork case; the rest of the digest changes are a performance booster on their own.

**Comment by cotequeiroz** [comment]
Also, please consider that the digest changes boost performance as well.  They will not be shown in `openssl speed` comparisons, as they enhance the case where there are multiple updates, and `speed` does init, calls update only once, and then final.  However, this gives an idea of how much of an improvement a single 8192 (3 iocts - open, 1*update, final) operation does in comparison with calling 8 updates of 1024 bytes (10 ioctls), for example:
```
openssl speed -elapsed -evp sha256 2> /dev/null
OpenSSL 1.1.1c  28 May 2019
built on: Tue May 28 20:22:13 2019 UTC
options:bn(64,32) rc4(char) des(long) aes(partial) idea(int) blowfish(ptr)
compiler: arm-openwrt-linux-muslgnueabi-gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -O3 -fpic -ffunction-sections -fdata-sections -znow -zrelro -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DZLIB -DZLIB_SHARED -DNDEBUG -DOPENSSL_PREFER_CHACHA_OVER_GCM
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
sha256             844.44k     2257.26k     8775.77k    31247.70k   111899.99k   137603.75k
```
If you consider HMAC copies, by deferring initialization, you get 1 update ioctl avoided in the "parent" context, and 1 ioctl avoided per copy.

**Comment by cotequeiroz** [comment]
I've added tracers to the digest calls, and counted the number of `digest_op` calls, where the update/final ioctls are used, using `openssl s_client` to retrieve an URL (~14K data).
```
$ egrep -c digest_op ossl.dbg*
ossl.dbg.std:196
ossl.dbg.new:68
```

**Comment by levitte** [comment]
> They will not be shown in `openssl speed` comparisons, as they enhance the case where there are multiple updates, and `speed` does init, calls update only once, and then final.

Actually, it's possible to have `openssl speed` run multiple processes on Unix...

```
: ; openssl speed -help
Usage: speed [options] ciphers...
Valid options are:
 -help               Display this summary
 -evp val            Use EVP-named cipher or digest
 -decrypt            Time decryption instead of encryption (only EVP)
 -aead               Benchmark EVP-named AEAD cipher in TLS-like sequence
 -mb                 Enable (tls1>=1) multi-block mode on EVP-named cipher
 -mr                 Produce machine readable output
 -multi +int         Run benchmarks in parallel
 -async_jobs +int    Enable async mode and start specified number of jobs
 -rand val           Load the file(s) into the random number generator
 -writerand outfile  Write random data to the specified file
 -engine val         Use engine, possibly a hardware device
 -elapsed            Use wall-clock time instead of CPU user time as divisor
 -primes +int        Specify number of primes (for RSA only)
 -seconds +int       Run benchmarks for specified amount of seconds
 -bytes +int         Run [non-PKI] benchmarks on custom-sized buffer
 -misalign +int      Use specified offset to mis-align buffers
```

Notice the `-multi` option...

**Comment by cotequeiroz** [comment]
It’s not a matter of multitasking, but of calling multiple updates using the same CTX. Correct me if I’m wrong, but won’t running ‘-multi 5’ just use 5 CTXs in parallel?

**Comment by levitte** [comment]
Yes. Sorry, misread what you had written

**Comment by levitte** [comment]
I lost track of this, sorry.  Where do we stand?

**Comment by cotequeiroz** [comment]
I shall probably rewrite the subject & goal of this PR. It's more of a speed up feature than a fix/workaround for a problem.  It has a side-effect that kind of works around the problem with fork.

In my (not really knowledgeable) view he problem with forks & open contexts in kernel won't be easily fixed.  As long as engines are transparent to the user of the library, the expectation is that it should work the same way as the openssl/default engine.  By design or not, the default implementation allows the use of all open contexts in both processes after a fork, and many applications take advantage of this already.

When using engines that keep an open context at the kernel level, copying the memory structure won't be enough.  There has to be a mechanism to duplicate those contexts before they are closed/used again.  Is there any way to detect a fork from openssl, so that the contexts can be duplicated?  A callback when a fork is called would work.  I don't know any way to do it.  A check, by the engine, if the pid has changed (or any other way for the library to detect a fork from the parent process) with every crypto call would almost work--as long as the parent process does not close/update any context before every child has had the opportunity to clone them.  Another solution would be to add a call (or option to `OPENSSL_init_crypto`) to be made after the fork to copy the contexts: the child would clone all contexts, while the parent would wait for the child to finish it. 

In conclusion, I can't see a good solution to this.  So, as it is, the PR merits are more towards improving hashing performance than making it work across forks.  Allow me to update the title and description to show that and avoid confusion.  I'll strike through the description to maintain the history.

**Comment by richsalz** [comment]
Look at doc/man3/OPENSSL_fork_prepare.pod

**Comment by cotequeiroz** [comment]
Great!  I'll keep the subject and description then.  We need to setup hooks so that the engine can duplicate the contexts.  If you know a way to do it, let me know; meanwhile, I'll study those calls to see what can be done.  Thanks.

**Comment by t8m** [comment]
I am going to close this as engines are legacy and supporting forks in e_devcrypto would be really a new feature of the engine.
